### PR TITLE
Email Service

### PR DESCRIPTION
**Pull request by sirajudheenam**
@ArtieReus, @edda. Please review.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -75,7 +75,7 @@ gem 'arc-client', git: 'https://github.com/sapcc/arc-client.git'
 gem 'sdoc', '~> 1.1.0', group: :doc
 
 # if you update puma check Dockerfile for Version dependencies!
-gem 'puma', '= 4.3.9', require: false
+gem 'puma', '= 3.12.6', require: false
```
Comment (ArtieReus) [inline Gemfile:78]: Did you notice that you downgrade puma? 😅
Please just add your required libraries without downgrade or upgrade the rest of the libraries.

File: Gemfile.lock [hunk 2]
```diff
@@ -349,7 +362,7 @@ GEM
       guard (~> 2.1)
       guard-compat (~> 1.1)
       rspec (>= 2.99.0, < 4.0)
-    haml (5.2.2)
+    haml (5.0.2)
```
Comment (ArtieReus) [inline Gemfile.lock:365]: Hier the same as before. Do not change versions please.

File: Gemfile.lock [hunk 3]
```diff
@@ -407,7 +421,7 @@ GEM
       mime-types-data (~> 3.2015)
     mime-types-data (3.2019.0331)
     mini_mime (1.0.3)
-    mini_portile2 (2.6.1)
+    mini_portile2 (2.4.0)
```
Comment (ArtieReus) [inline Gemfile.lock:424]: same same

**Review by ArtieReus — CHANGES_REQUESTED** [review]
Please just add required libraries for the email sevice without downgrading or upgrading the rest of the libraries. THX!!

File: plugins/email_service/config/policy.json [hunk 4]
```diff
@@ -1,3 +1,30 @@
 {
-  "email_service:application_list": "@"
+  "email_service_user": "role:email_user",
+  "email_service_admin": "role:email_admin",
+  "email_role_required": "role:email_user or role:email_admin",
```
Comment (ArtieReus) [inline plugins/email_service/config/policy.json:4]: As far I can see there is no difference for any of the actions described here between email service user oder email admin? Is this right? Why do you have then two roles?

File: plugins/email_service/config/routes.rb [hunk 5]
```diff
@@ -1,3 +1,44 @@
 EmailService::Engine.routes.draw do
-  get '/' => 'application#index', as: :index
+
+  get '/' => 'web#index', as: :index
+
+  get 'emails/stats' => 'emails#stats'
```
Comment (ArtieReus) [inline plugins/email_service/config/routes.rb:5]: I didn't found any policy for this route.

File: plugins/email_service/config/routes.rb [hunk 6]
```diff
@@ -1,3 +1,44 @@
...
+  get '/emails/info' => 'emails#info'
```
Comment (ArtieReus) [inline plugins/email_service/config/routes.rb:6]: I didn't found any policy for this route.

File: plugins/email_service/config/routes.rb [hunk 7]
```diff
@@ -1,3 +1,44 @@
...
+
+  get 'settings/show_config' => 'settings#show_config'
+  get 'settings' => 'settings#index'
+  post 'settings/enable_cronus' => 'settings#enable_cronus'
```
Comment (ArtieReus) [inline plugins/email_service/config/routes.rb:10]: I didn't found any policy for this route.

File: plugins/email_service/config/routes.rb [hunk 8]
```diff
@@ -1,3 +1,44 @@
...
+  post 'settings/disable_cronus' => 'settings#disable_cronus'
```
Comment (ArtieReus) [inline plugins/email_service/config/routes.rb:11]: I didn't found any policy for this route.

File: plugins/email_service/config/routes.rb [hunk 9]
```diff
@@ -1,3 +1,44 @@
...
+  
+  get 'configset' => 'configset#index'
+  get 'configset/new_configset' => 'configset#new_configset'
```
Comment (ArtieReus) [inline plugins/email_service/config/routes.rb:14]: I didn't found any policy for this route.

File: plugins/email_service/config/routes.rb [hunk 10]
```diff
@@ -1,3 +1,44 @@
...
+  get 'configset/show_configset' => 'configset#show_configset'
+
+  post 'configset/create_configset' => 'configset#create_configset'
```
Comment (ArtieReus) [inline plugins/email_service/config/routes.rb:17]: I didn't found any policy for this route.

File: plugins/email_service/config/routes.rb [hunk 11]
```diff
@@ -1,3 +1,44 @@
...
+  post 'configset/destroy_configset' => 'configset#destroy_configset'
```
Comment (ArtieReus) [inline plugins/email_service/config/routes.rb:18]: I didn't found any policy for this route.

File: plugins/email_service/config/routes.rb [hunk 12]
```diff
@@ -1,3 +1,44 @@
...
+    end
+  end
+
+ 
+  resources :verifications, only: [:index, :show, :new, :create, :destroy]
+  resources :domain_verifications
+
+  post 'verify_dkim', action: :verify_dkim, controller: 'domain_verifications'
+  post 'activate_dkim', action: :activate_dkim, controller: 'domain_verifications'
+  post 'deactivate_dkim', action: :deactivate_dkim, controller: 'domain_verifications'
```
Comment (ArtieReus) [inline plugins/email_service/config/routes.rb:42]: I would check if all actions are in the policy.

File: plugins/email_service/config/policy.json (hunk 4 above)
Comment (sirajudheenam) [reply plugins/email_service/config/policy.json:4]: Disabling and enabling the cronus require `email_admin` role. All other functionalities merely require `email_user` role.

File: plugins/email_service/app/services/service_layer/email_service_services/email.rb [hunk 13]
```diff
@@ -0,0 +1,23 @@
+# frozen_String_literal: true
+
+module ServiceLayer
+  module EmailServiceServices
```
[the remark this answers is not in the document]
Comment (sirajudheenam) [reply plugins/email_service/app/services/service_layer/email_service_services/email.rb:4]: Removed all together since I have defined the needed methods under Identity Service AWS module.

File: plugins/email_service/app/services/service_layer/email_service_service.rb [hunk 14]
```diff
@@ -3,12 +3,26 @@
...
+    include EmailServiceServices::VerifiedEmail
+
     def available?(_action_name_sym = nil)
-      elektron.service?('cronus')
+      elektron.service?('email-aws')
+    end
+
+    def elektron_email_service
+      @elektron_email_service ||= elektron.service(
+        'identity', path_prefix: '/v3'
```
[the remark this answers is not in the document]
Comment (sirajudheenam) [reply plugins/email_service/app/services/service_layer/email_service_service.rb:16]: Added under Identity Service AWS module.

File: plugins/email_service/config/routes.rb (hunk 5 above)
Comment (sirajudheenam) [reply plugins/email_service/config/routes.rb:5]: added policies.

File: plugins/email_service/config/routes.rb (hunk 6 above)
Comment (sirajudheenam) [reply plugins/email_service/config/routes.rb:6]: added policies.

File: config/navigations/services_navigation.rb [hunk 15]
```diff
@@ -286,11 +286,11 @@
                  'Services',
                  nil,
                  html: { class: 'fancy-nav-header', 'data-icon': 'service-icon' },
-                 if: -> { services.available?(:email_service) } do |services_nav|
+                 if: -> { plugin_available?(:email_service) } do |services_nav|
```
Comment (ArtieReus) [inline config/navigations/services_navigation.rb:289]: why did you changed the `services.available?` check? I would check if the service `cronus` is available and as a fall back check if the plugin exists. See here:
https://github.com/sapcc/elektra/blob/81a55a49aba95b8c6fc85e03c021f44fc5c48b02/config/navigations/services_navigation.rb#L64
Afterwards you should define the `available?` method in your service layer.
https://github.com/sapcc/elektra/blob/81a55a49aba95b8c6fc85e03c021f44fc5c48b02/plugins/compute/app/services/service_layer/compute_service.rb#L21

File: config/navigations/services_navigation.rb [hunk 16]
```diff
@@ -286,11 +286,11 @@
...
       services_nav.item :email_service,
-                        'Email',
+                        'eMail',
                         -> { plugin('email_service').index_path },
-                        if: -> { services.available?(:email_service) },
+                        if: -> { plugin_available?(:email_service) },
```
Comment (ArtieReus) [inline config/navigations/services_navigation.rb:293]: The same as before here.

File: plugins/email_service/app/controllers/email_service/application_controller.rb [hunk 17]
```diff
@@ -2,11 +2,55 @@
...
+      end
     end
+ 
+    def restrict_access
+      unless current_user.has_role?('cloud_support_tools_viewer')
+        redirect_to index_path
+      end
+    end 
+
+    def render_paginatable_for_configset(items, filter={})
```
Comment (ArtieReus) [inline plugins/email_service/app/controllers/email_service/application_controller.rb:23]: Why do you have this method globally when it is just for the configset. Why not in the configset controller? You don't have to changed it is just best practices.

File: plugins/email_service/app/controllers/email_service/configset_controller.rb [hunk 18]
```diff
@@ -0,0 +1,77 @@
+module EmailService
+  class ConfigsetController < ::EmailService::ApplicationController
+    before_action :restrict_access
```
Comment (ArtieReus) [inline plugins/email_service/app/controllers/email_service/configset_controller.rb:3]: How you enforce your policy? I couldn't find any reference to the 
```
authorization_context 'email_service'
authorization_required
```

File: plugins/email_service/app/controllers/email_service/domain_verifications_controller.rb [hunk 19]
```diff
@@ -0,0 +1,138 @@
+module EmailService
+  class DomainVerificationsController < ::EmailService::ApplicationController
+    before_action :restrict_access
```
Comment (ArtieReus) [inline plugins/email_service/app/controllers/email_service/domain_verifications_controller.rb:3]: How you enforce your policy? I couldn't find any reference to the 
```
authorization_context 'email_service'
authorization_required
```

File: plugins/email_service/app/controllers/email_service/emails_controller.rb [hunk 20]
```diff
@@ -0,0 +1,112 @@
+module EmailService
+  class EmailsController < ::EmailService::ApplicationController
+    before_action :restrict_access
```
Comment (ArtieReus) [inline plugins/email_service/app/controllers/email_service/emails_controller.rb:3]: How you enforce your policy? I couldn't find any reference to the 
```
authorization_context 'email_service'
authorization_required
```

File: plugins/email_service/app/controllers/email_service/settings_controller.rb [hunk 21]
```diff
@@ -0,0 +1,41 @@
+module EmailService
+  class SettingsController < ::EmailService::ApplicationController
+    before_action :restrict_access
```
Comment (ArtieReus) [inline plugins/email_service/app/controllers/email_service/settings_controller.rb:3]: How you enforce your policy? I couldn't find any reference to the 
```
authorization_context 'email_service'
authorization_required
```

File: plugins/email_service/app/controllers/email_service/templated_emails_controller.rb [hunk 22]
```diff
@@ -0,0 +1,91 @@
+module EmailService
+  class TemplatedEmailsController < ::EmailService::ApplicationController
+    before_action :restrict_access
```
Comment (ArtieReus) [inline plugins/email_service/app/controllers/email_service/templated_emails_controller.rb:3]: How you enforce your policy? I couldn't find any reference to the 
```
authorization_context 'email_service'
authorization_required
```

File: plugins/email_service/app/controllers/email_service/templates_controller.rb [hunk 23]
```diff
@@ -0,0 +1,113 @@
+module EmailService
+  class TemplatesController < ::EmailService::ApplicationController
+    before_action :restrict_access
```
Comment (ArtieReus) [inline plugins/email_service/app/controllers/email_service/templates_controller.rb:3]: How you enforce your policy? I couldn't find any reference to the 
```
authorization_context 'email_service'
authorization_required
```

File: plugins/email_service/app/controllers/email_service/verifications_controller.rb [hunk 24]
```diff
@@ -0,0 +1,80 @@
+module EmailService
+  class VerificationsController < ::EmailService::ApplicationController
+    before_action :restrict_access
```
Comment (ArtieReus) [inline plugins/email_service/app/controllers/email_service/verifications_controller.rb:3]: How you enforce your policy? I couldn't find any reference to the 
```
authorization_context 'email_service'
authorization_required
```

File: plugins/email_service/app/controllers/email_service/web_controller.rb [hunk 25]
```diff
@@ -0,0 +1,10 @@
+module EmailService
+  class WebController < ::EmailService::ApplicationController
```
Comment (ArtieReus) [inline plugins/email_service/app/controllers/email_service/web_controller.rb:2]: How you enforce your policy? I couldn't find any reference to the 
```
authorization_context 'email_service'
authorization_required
```

File: plugins/email_service/app/services/service_layer/email_service_service.rb [hunk 26]
```diff
@@ -1,14 +0,0 @@
-# frozen_string_literal: true
-
-module ServiceLayer
```
Comment (ArtieReus) [inline plugins/email_service/app/services/service_layer/email_service_service.rb:3]: Why did you remove the email service? You don't need Cronus anymore?

File: plugins/identity/app/services/service_layer/identity_service.rb [hunk 27]
```diff
@@ -9,6 +9,9 @@ class IdentityService < Core::ServiceLayer::Service
     include IdentityServices::Group
     include IdentityServices::Role
     include IdentityServices::RoleAssignment
+    include IdentityServices::AWS::Ec2Creds
```
Comment (ArtieReus) [inline plugins/identity/app/services/service_layer/identity_service.rb:12]: I think you didn't get my point right yesterday. Sorry... Everything related to `Openstack Identity Service` should go to the identity service layer. And if there is any method missing from the identity api you should then do it in the identity service layer. You can fetch the identity service layer from any controller in your plugin.

These AWS modules are very specific and just for your plugin. Please move this modules back to the email_service layer. If it is not clear give me call. Thx!!

**Review by ArtieReus — CHANGES_REQUESTED** [review]

File: plugins/email_service/app/controllers/email_service/domain_verifications_controller.rb (hunk 19 above)
Comment (sirajudheenam) [reply plugins/email_service/app/controllers/email_service/domain_verifications_controller.rb:3]: done.

File: plugins/email_service/app/controllers/email_service/configset_controller.rb (hunk 18 above)
Comment (sirajudheenam) [reply plugins/email_service/app/controllers/email_service/configset_controller.rb:3]: done.

File: config/navigations/services_navigation.rb (hunk 16 above)
Comment (sirajudheenam) [reply config/navigations/services_navigation.rb]: done.

File: config/navigations/services_navigation.rb (hunk 15 above)
Comment (sirajudheenam) [reply config/navigations/services_navigation.rb]: done.

File: plugins/email_service/app/controllers/email_service/application_controller.rb (hunk 17 above)
Comment (sirajudheenam) [reply plugins/email_service/app/controllers/email_service/application_controller.rb]: done.

File: plugins/email_service/app/controllers/email_service/web_controller.rb (hunk 25 above)
Comment (sirajudheenam) [reply plugins/email_service/app/controllers/email_service/web_controller.rb:2]: done.

File: plugins/email_service/app/services/service_layer/email_service_service.rb (hunk 26 above)
Comment (sirajudheenam) [reply plugins/email_service/app/services/service_layer/email_service_service.rb:3]: restored.

File: plugins/identity/app/services/service_layer/identity_service.rb (hunk 27 above)
Comment (sirajudheenam) [reply plugins/identity/app/services/service_layer/identity_service.rb]: moved back as before.

File: plugins/email_service/app/services/service_layer/email_service_services/email.rb [hunk 28]
```diff
@@ -0,0 +1,23 @@
...
+    # email api implementation
+    module Email
+
+      # Creds = Struct.new(:access, :secret)
+      Creds = Struct.new(:access, :secret, :error)
+      def aws_creds(user_id, filter = {})
+
+        response = elektron_email_service.get("users/#{user_id}/credentials/OS-EC2")
```
Comment (ArtieReus) [inline plugins/email_service/app/services/service_layer/email_service_services/email.rb:12]: As I mentioned you should use the methods already defined in Identity. Do not do it from your plugin. See hier:
https://github.com/sapcc/elektra/blob/master/plugins/identity/app/services/service_layer/identity_service.rb#L17
If you need such credentials you should define a method here:
https://github.com/sapcc/elektra/blob/master/plugins/identity/app/services/service_layer/identity_services/user.rb
and called from this controller like
`services.identity.credentials("OS_EC2")`

**Review by ArtieReus — CHANGES_REQUESTED** [review]

File: plugins/email_service/app/controllers/email_service/configset_controller.rb [hunk 29]
```diff
@@ -0,0 +1,111 @@
...
+    def show_configset
+      name = params[:name] if params[:name]
+      @configset_description = describe_configset(name)
+    rescue Elektron::Errors::ApiResponse => e
+      flash[:error] = "Status Code: #{e.code} : Error: #{e.message}"
+    rescue Exception => e
+      flash[:error] = "Status Code: 500 : Error: #{e.message}"
+    end
+
+    def create_configset
```
[the remark this answers is not in the document]
Comment (sirajudheenam) [reply plugins/email_service/app/controllers/email_service/configset_controller.rb:36]: Audit Logging is done at Cronus API level. When UI makes calls to Cronus API, it in turn writes the events. Do I still need to implement from UI?

File: plugins/email_service/app/services/service_layer/email_service_service.rb [hunk 30]
```diff
@@ -3,12 +3,25 @@
...
+    # include EmailServiceServices::VerifiedEmail
+
     def available?(_action_name_sym = nil)
-      elektron.service?('cronus')
+      elektron.service?('email-aws')
     end
 
-    def test
-      elektron.service('email_service').get('/')
+    def elektron_email_service
```
[the remark this answers is not in the document]
Comment (sirajudheenam) [reply plugins/email_service/app/services/service_layer/email_service_service.rb]: Added credential method at Identity Service. Removed Email Service.

File: plugins/email_service/app/services/service_layer/email_service_services/email.rb (hunk 28 above)
Comment (sirajudheenam) [reply plugins/email_service/app/services/service_layer/email_service_services/email.rb]: done.

File: config/navigations/services_navigation.rb (hunk 16 above)
Comment (sirajudheenam) [reply config/navigations/services_navigation.rb]: I need to validate using `plugin_available?` method now as there is no `email_service` registered now.

File: plugins/email_service/app/controllers/email_service/configset_controller.rb (hunk 29 above)
[the remark this answers is not in the document]
Comment (ArtieReus) [reply plugins/email_service/app/controllers/email_service/configset_controller.rb:36]: I would say yes please. Everything triggered from elektra needs also an own log entry. You would never now if the triggered action in cronus is directly per API or through elektra.

File: plugins/identity/app/services/service_layer/identity_services/credential.rb [hunk 31]
```diff
@@ -0,0 +1,27 @@
...
+    # This module implements Openstack User API
+    module Credential
+
+      Creds = Struct.new(:access, :secret, :error)
+
+      def credential_map
+        @pcredential_map ||= class_map_proc(Identity::Credential)
+      end
+
+      def credentials(user_id, filter = {})
```
Comment (ArtieReus) [inline plugins/identity/app/services/service_layer/identity_services/credential.rb:14]: This is just returning OS-EC2 credentials, why not calling the method `aws_credentials`? this would not lead to the confusion of which kind of credentials are being returned. Bdw. I couldn't find the method in the identity API documentation https://docs.openstack.org/api-ref/identity/v3/index.html#credentials, I would imagine that the API call your are using it is already removed from the latest version.

**Review by ArtieReus — CHANGES_REQUESTED** [review]

File: config/navigations/services_navigation.rb (hunk 16 above)
Comment (ArtieReus) [reply config/navigations/services_navigation.rb]: this just check if there is a plugin called like this and the plugin exists already. Keep it.

File: plugins/email_service/app/controllers/email_service/configset_controller.rb (hunk 29 above)
[the remark this answers is not in the document]
Comment (sirajudheenam) [reply plugins/email_service/app/controllers/email_service/configset_controller.rb:36]: Audit logging is added for CRUD functions.

File: plugins/identity/app/services/service_layer/identity_services/credential.rb (hunk 31 above)
Comment (sirajudheenam) [reply plugins/identity/app/services/service_layer/identity_services/credential.rb]: changed the name to `AWSCredential` / `aws_credentials`

File: plugins/email_service/app/controllers/email_service/configset_controller.rb (hunk 29 above)
[the remark this answers is not in the document]
Comment (ArtieReus) [reply plugins/email_service/app/controllers/email_service/configset_controller.rb:36]: Thx! I have found the audit_logger entries. Doesn't this configset_controller also the audit_logger?

File: plugins/identity/app/services/service_layer/identity_service.rb [hunk 32]
```diff
@@ -9,6 +9,7 @@ class IdentityService < Core::ServiceLayer::Service
     include IdentityServices::Group
     include IdentityServices::Role
     include IdentityServices::RoleAssignment
+    include IdentityServices::AWSCredential
```
Comment (ArtieReus) [inline plugins/identity/app/services/service_layer/identity_service.rb:12]: Please leave the identity service name as before `Credential`. If we need in the future to add more credentials this module would not fit anymore. Changing the method name is enough. Thx!

**Review by ArtieReus — CHANGES_REQUESTED** [review]
looks good. Just rename the identity modul back to Credentials and check if we need a looger for the configset controller :)

File: plugins/identity/app/services/service_layer/identity_service.rb (hunk 32 above)
Comment (sirajudheenam) [reply plugins/identity/app/services/service_layer/identity_service.rb]: done.

File: plugins/identity/app/models/identity/aws_credential.rb [hunk 33]
```diff
@@ -0,0 +1,5 @@
+module Identity
+  class AWSCredential < Core::ServiceLayer::Model
```
Comment (ArtieReus) [inline plugins/identity/app/models/identity/aws_credential.rb:2]: This should be also changed to Credential or?

**Review by ArtieReus — CHANGES_REQUESTED** [review]

File: plugins/identity/app/models/identity/aws_credential.rb (hunk 33 above)
Comment (sirajudheenam) [reply plugins/identity/app/models/identity/aws_credential.rb:2]: Let us keep it this way because it is specific to AWS/EC2.

Comment (ArtieReus) [reply plugins/identity/app/models/identity/aws_credential.rb:2]: Short question… where are you actually using the model AWSCredential??
in your aws_credentials you are not mapping to the model?
https://github.com/sapcc/elektra/pull/1013/files#diff-828919a30a0401afc91327539613ecd306bc81a20a0795c36149d9fdc7875aefR14
See an example hier:
https://github.com/sapcc/elektra/blob/master/plugins/identity/app/services/service_layer/identity_services/project.rb#L16

Comment (ArtieReus) [reply plugins/identity/app/models/identity/aws_credential.rb:2]: If no needed just remove the model.

**Review by ArtieReus — CHANGES_REQUESTED** [review]

**Review by ArtieReus — APPROVED** [review]
lgtm
